processor/processor.go

```diff
@@ -26,10 +26,62 @@ func BasicSpanName(receiver, function string) string {
 	return receiver + "." + function
 }
 
-// Processor traverses AST, collects details on funtions and methods, and invokes Instrumenter
+func functionLiteral(fnc *ast.FuncDecl, funcLitInst *funcTypeConditions) *ast.FuncLit {
+	if fnc == nil || fnc.Body == nil || len(fnc.Body.List) != 1 {
+		return nil
+	}
+	if funcLitInst == nil || !funcLitInst.hasContext {
+		return nil
+	}
+
+	returnStmt, stmtOk := fnc.Body.List[0].(*ast.ReturnStmt)
+	if !stmtOk || len(returnStmt.Results) != 1 {
+		return nil
+	}
+
+	funcLit, funcLitOk := returnStmt.Results[0].(*ast.FuncLit)
+	if !funcLitOk || funcLit.Body == nil {
+		return nil
+	}
+	funcLitInst = nil
+
+	return funcLit
+}
+
+// funcTypeConditions collects details on functions and methods
+type funcTypeConditions struct {
+	Type       *ast.FuncType
+	hasContext bool
+	hasError   bool
+}
+
+func (p *Processor) hasPrefixConditions(fn *funcTypeConditions) {
+	if t := fn.Type; t != nil {
+		if ps := t.Params; ps != nil {
+			for _, q := range ps.List {
+				if q == nil {
+					continue
+				}
+				fn.hasContext = fn.hasContext || p.isContext(*q)
+			}
+		}
+
+		if rs := t.Results; rs != nil {
+			for _, q := range rs.List {
+				if q == nil {
+					continue
+				}
+				fn.hasError = fn.hasError || p.isError(*q)
+			}
+		}
+	}
+}
```

Comment:
I guess this is two separate functions. + return early from ifs when main chunk of body.
```suggestion
func (p *Processor) funcitonHasContext(fn *ast.FuncType) bool {
	t := fn.Type
	if t == nil {
	   return false
	}
        ps := t.Params
        if ps == nil {
           return false
        }
	for _, q := range ps.List {
		if q == nil {
			continue
		}
		return p.isContext(*q)
	}
	return false
}

...
```

Reply:
I'll separate these.